static/gallery/js/infinite_scroll.js

```diff
@@ -28,31 +32,25 @@ function load_images_from_cursor(cursor) {
     // Find the images in the thumbnail container
     var images = document.querySelectorAll('.image');
     var thumbnails = document.querySelectorAll('.thumbnail');
-    // Ignore pages with no images
-    if (images == []) {
-        return;
-    }
-
-    // Change the source of the next page of images
-    for (var i=scroll_cursor; i < scroll_cursor + pagination_size; i++) {
-
-        if (i >= images.length) {
-            // All images have been set to load
-            window.removeEventListener('scroll',check_infinite_scroll);
-            return;
-        }
 
-        // Re-justify once the data has loaded
+    var i = scroll_cursor;      // start from last position
+    
+    // Change the source of the next page of images if any left
+    for (; i < images.length && i < scroll_cursor + pagination_size; i++) {
+        // Re-justify after this image has loaded
         images[i].addEventListener('load', justify_images);
-        // Show the image as they load hidden
+        // and change its display property to show it
         thumbnails[i].style.display = 'block';
+        // Initiate the load
         var src = images[i].getAttribute('data-src');
         images[i].setAttribute('src', src);
     }
-
-    scroll_cursor += pagination_size;
-
+    if (i >= images.length && scroll_cursor < images.length) {
```

Comment:
Not sure I understand the logic of this. Could `i` not be kept inside the scope of the for loop and then simply `if (scroll_cursor == images.length)` (after restoring the line to increment it by the `pagination_size`)

Reply:
Of course it could.  The important part is to update scroll_cursor to images.length before the "return" in the previous version - this was the defect.

This last version I submitted was just a design-style. The 'state-variables' of this problem as I can see are `scroll_cursor`, the actual state of the individual `image` elements (not loading or loading), and the queue of scroll events...

The previous versions internal `if `statement detected that the last segment is being loaded, initiated their load and left the loop and the function "in a hurry".   Even after the fix of setting scroll_cursor to image.length in that inner `if`,  subsequent scroll events were coming back and going through 1) checking images length again against 0,  2)  going into the for loop,  3) falling into the if again  4) repeating the window removeEventListener, and 5) hopefully setting the scroll_cursor to image_length and not i...

The way I structured it there are only two 'focused' condition checks instead of four 'distributed' ones above:
1) the `for` loop condition, and that is the accurate condition under which the for loop body should be entered and executed;
2) After the for loop the `if` detects the state transition from 'not-all-images-started-loading' to the 'all-images-loading' state (nothing more to do under infinite-scroll), and this state transition removes the listener once only (the other solution required yet another check to avoid repeated listener removals if scroll events have already been queued for this listener.)

Not to be overly pedantic, both solutions do the same, which one is easier to understand depends on the reader...  The single function solution would be just an 'elegance' and conciseness exercise, tbd...